bindings/nodejs/index.js

```diff
@@ -19,9 +19,59 @@
 
 /// <reference types="node" />
 
+const { Writable, Readable } = require('node:stream')
+
+class ReaderStream extends Readable {
```

Comment:
We need to give more consideration to this API.

## For naming

Do we still `Reader` and `Writer` for users? How about rename to `RawReader` / `RawWriter` (or other things) and expose `class Reader extends Readable` and `class Writer extends Writable`?

In this, our users can write code like:

```js
let stream = Readable.from(op.readerSync());
```

or 

```js
op.readerSync().pipe(abc);
```

## For async

Do `Readable` and `Writable` have async support? How could we support them?

Reply:
I'd keep the reader and writer to preserve the possibility for users to make their own calls.

> let stream = Readable.from(op.readerSync());
> op.readerSync().pipe(abc);

We can't produce a stream directly. There is no such usage as `from`.

> Do `Readable` and `Writable` have async support? How could we support them?

The stream of nodejs is asynchronous. This is an event loop model.

Reply:
> We can't produce a stream directly. There is also no such usage as `from`.

I'm talking about this API: https://nodejs.org/api/stream.html#streamreadablefromiterable-options


> I'd keep the reader and writer to preserve the possibility for users to make their own calls.

Make sense. How about following the same pattern like `node::fs` does?

We can have API:

```
let r = op.reader() -> Reader

r.read(buffer)
r.createReadStream() -> ReadStream
```

Reply:
> The stream of nodejs is asynchronous. This is an event loop model.

I didn't get this, can you explain a lit? It looks like `ReadStream` only calls `readerSync`.

Reply:
> I'm talking about this API: https://nodejs.org/api/stream.html#streamreadablefromiterable-options

I know what you mean. However, it requires that the parameter be implemented as an iterator. 
We can't implement that at the moment, `napi-rs` haven't implemented this.


> Make sense. How about following the same pattern like `node::fs` does?

It looks better. But we can't add instance methods to these exported classes, so it can't be implemented.

Reply:
Can I have a try on this?

Reply:
Maybe you could take this as a reference. https://nodejs.org/en/docs/guides/event-loop-timers-and-nexttick

The nodejs only has one main thread. So it takes the event loop as its async model. Every task is executed one by one.
Although we are calling readerSync, it's non-blocking through the event loop. The timing of the call is scheduled by the runtime.

Reply:
> The nodejs only has one main thread. So it takes the event loop as its async model. Every task is executed one by one. Although we are calling readerSync, it's non-blocking through the event loop. The timing of the call is scheduled by the runtime.

The point is not about `readerSync`. I'm wondering why we don't have `BlokcingReadStream` and `ReadSteam` to match with `BlockingReader` and `Reader`.

Reply:
> The point is not about `readerSync`. I'm wondering why we don't have `BlokcingReadStream` and `ReadSteam` to match with `BlockingReader` and `Reader`.

It's okay to add them. I'll add.

Reply:
> It's okay to add them. I'll add.

The main question here is **why**. I'm not seeing this is wrong. Are there other reasons you chose not to include an async version?

Reply:
> Can we implement this way:

Well, it is indeed a viable way. I'll add this.

Reply:
> I'm wondering why we don't have `BlokcingReadStream` and `ReadSteam` to match with `BlockingReader` and `Reader`.

I reconsidered. This one is no need to add. 
Their name can't really reflect blocking. the stream itself is not blocking.

Reply:
Node's `fshandle` support both async and sync operations in the same class but we have `Reader` and `BlockingReader` here.